src/devices/SocketCan/Interop.cs

```diff
@@ -0,0 +1,235 @@
+            {
+                int totalBytesWritten = 0;
+                while (totalBytesWritten < buffer.Length)
+                {
+                    int bytesWritten = Interop.SocketWrite((int)handle.DangerousGetHandle(), b, buffer.Length);
+                    if (bytesWritten < 0)
+                    {
+                        throw new IOException("`write` operation failed");
+                    }
+
+                    totalBytesWritten += bytesWritten;
+                }
+            }
+        }
+
+        public static unsafe int Read(SafeHandle handle, Span<byte> buffer)
+        {
+            fixed (byte* b = buffer)
+            {
+                int bytesRead = Interop.SocketRead((int)handle.DangerousGetHandle(), b, buffer.Length);
+                if (bytesRead < 0)
+                {
+                    throw new IOException("`read` operation failed");
+                }
+
+                return bytesRead;
+            }
+        }
+
+        public static void CloseSocket(IntPtr fd)
+        {
+            CloseSocket((int)fd);
+        }
+
+        public static IntPtr CreateCanRawSocket(string networkInterface)
+        {
+            const int SOCK_RAW = 3;
+            int socket = CreateNativeSocket(PF_CAN, SOCK_RAW, CanProtocol.CAN_RAW);
+
+            if (socket == -1)
+                throw new IOException("CAN socket could not be created");
+
+            BindToInterface(socket, networkInterface);
+
+            return new IntPtr(socket);
+        }
+
+        public static bool SetCanRawSocketOption<T>(SafeHandle handle, CanSocketOption optName, ReadOnlySpan<T> data)
+            where T : struct
+        {
+            return SetSocketOption(handle, SOL_CAN_RAW, optName, data);
+        }
+
+        private static unsafe bool SetSocketOption<T>(SafeHandle handle, int level, CanSocketOption optName, ReadOnlySpan<T> data)
+            where T : struct
+        {
+            int fd = (int)handle.DangerousGetHandle();
+            ReadOnlySpan<byte> buf = MemoryMarshal.AsBytes(data);
+            fixed (byte* pinned = buf)
+            {
+                return SetSocketOpt(fd, level, (int)optName, pinned, buf.Length) == 0;
+            }
+        }
+
+        private static unsafe void BindToInterface(int fd, string interfaceName)
+        {
+            int idx = GetInterfaceIndex(fd, interfaceName);
+            CanSocketAddress addr = new CanSocketAddress();
+            addr.can_family = PF_CAN;
+            addr.can_ifindex = idx;
+
+            if (-1 == BindSocket(fd, ref addr, (uint)Marshal.SizeOf<CanSocketAddress>()))
+            {
+                throw new IOException($"Cannot bind to socket to `{interfaceName}`");
+            }
+        }
+
+        private static unsafe int GetInterfaceIndex(int fd, string name)
+        {
+            const uint SIOCGIFINDEX = 0x8933;
+            const int MaxLen = ifreq.IFNAMSIZ - 1;
+
+            if (Encoding.ASCII.GetByteCount(name) >= MaxLen)
```

Comment:
>Encoding.ASCII.GetByteCount(name) [](start = 16, length = 33)

I believe you need to use the ASCII encoding which will throw on the invalid characters. right?

Reply:
actually I think this should be UTF8 (and yes it should throw) - will double check and fix

Reply:
leaving this ASCII until since we only expect name in format `can[0-9]`